include/deal.II/base/array_view.h

```diff
@@ -423,48 +425,180 @@ ArrayView<ElementType>::operator[](const std::size_t i) const
 
 
 /**
- * Create a view to an entire C-style array. This is equivalent to
- * initializing an ArrayView object with a pointer to the first element and
- * the size of the given argument.
+ * Create an ArrayView that takes a pair of iterators as arguments. The type
+ * of the ArrayView is inferred from the value type of the iterator (e.g., the
+ * view created from two const iterators will have a const type).
  *
- * Whether the resulting ArrayView is writable or not depends on the
- * ElementType being a const type or not.
+ * @warning The iterators @p begin and @p end must bound (in the usual half-open
+ * way) a contiguous in memory range of values. This function is intended for
+ * use with iterators into containers like
+ * <code>boost::container::small_vector</code> or <code>std::vector</code> and
+ * will not work correctly with, e.g.,
+ * <code>boost::container::stable_vector</code> or <code>std::deque</code>.
  *
- * @param[in] array The C-style array for which we want to have an ArrayView
- * object. The ArrayView corresponds to the <em>entire</em> vector.
+ * @relates ArrayView
+ */
+template <typename Iterator>
+ArrayView<typename std::remove_reference<typename std::iterator_traits<Iterator>::reference>::type>
+make_array_view (const Iterator begin, const Iterator end)
+{
+  static_assert(std::is_same<typename std::iterator_traits<Iterator>::iterator_category,
+                typename std::random_access_iterator_tag>::value,
+                "The provided iterator should be a random access iterator.");
+  Assert(begin <= end,
+         ExcMessage("The beginning of the array view should be before the end."));
+  // the reference type, not the value type, knows the constness of the iterator
+  return ArrayView<typename std::remove_reference
+         <typename std::iterator_traits<Iterator>::reference>::type>
+         (&*begin, end - begin);
```

Comment:
isn't there something like `std::address(*begin)` now?

Reply:
This was also not part of this PR but I'll address it anyway.